sapi/fpm/fpm/fpm_scoreboard.c

```diff
@@ -210,6 +210,64 @@ void fpm_scoreboard_release(struct fpm_scoreboard_s *scoreboard) {
 	scoreboard->lock = 0;
 }
 
+struct fpm_scoreboard_s *fpm_scoreboard_copy(struct fpm_scoreboard_s *scoreboard, int copy_procs)
+{
+	struct fpm_scoreboard_s *scoreboard_copy;
+	struct fpm_scoreboard_proc_s *scoreboard_proc_p;
+	size_t scoreboard_size, scoreboard_nprocs_size;
+	int i;
+	void *mem;
+
+	if (!scoreboard){
+		scoreboard = fpm_scoreboard_get();
+	}
+
+	if (copy_procs){
+		scoreboard_size        = sizeof(struct fpm_scoreboard_s) + (scoreboard->nprocs) * sizeof(struct fpm_scoreboard_proc_s *);
+		scoreboard_nprocs_size = sizeof(struct fpm_scoreboard_proc_s) * scoreboard->nprocs;
+
+		mem = emalloc(scoreboard_size + scoreboard_nprocs_size);
+	} else {
+		mem = emalloc(sizeof(struct fpm_scoreboard_s));
+	}
+
+	if (!mem) {
+		zlog(ZLOG_ERROR, "scoreboard: failed to allocate memory for copy");
+		return NULL;
+	}
+
+	scoreboard_copy = mem;
+
+	scoreboard = fpm_scoreboard_acquire(scoreboard, FPM_SCOREBOARD_LOCK_NOHANG);
+	if (!scoreboard){
+		efree(mem);
+		zlog(ZLOG_ERROR, "scoreboard: failed to lock (already locked)");
+		return NULL;
+	}
+
+	*scoreboard_copy = *scoreboard;
+
+	if (copy_procs) {
+		mem += scoreboard_size;
```

Comment:
You can just set `scoreboard_copy->procs = mem + scoreboard_size` and then you don't need to increment and set mem for each procs below.

Reply:
it means something like this should do (not tested):
```
scoreboard_copy->procs = mem + scoreboard_size;
for (i = 0; i < scoreboard->nprocs; i++) {
	scoreboard_proc_p = fpm_scoreboard_proc_acquire(scoreboard, i, FPM_SCOREBOARD_LOCK_HANG);
	*scoreboard_copy->procs[i] = *scoreboard_proc_p;
	fpm_scoreboard_proc_release(scoreboard_proc_p);
}
```

Reply:
Assuming my assumption from https://github.com/php/php-src/pull/3188#discussion_r177661662 is correct, but following your advice would make result in something like this (untested): 
```
scoreboard_copy_proc_p = mem + scoreboard_size;

for (i = 0; i < scoreboard->nprocs; i++) {
	scoreboard_copy->procs[i] = &scoreboard_copy_proc_p[i];
	scoreboard_proc_p = fpm_scoreboard_proc_acquire(scoreboard, i, FPM_SCOREBOARD_LOCK_HANG);
	*scoreboard_copy->procs[i] = *scoreboard_proc_p;
	fpm_scoreboard_proc_release(scoreboard_proc_p);
}
```
Is that what you had in mind?

Reply:
Missed that it's array of pointers so I think it's fine as it is now. Just ignore my comment :)